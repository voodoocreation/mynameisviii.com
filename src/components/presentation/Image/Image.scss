@import "../../../scss/variables", "../../../scss/mixins";

.Image {
  img {
    opacity: 1;
    transition: opacity 600ms ease-out;
  }

  &--wrapper {
    width: 100%;
  }

  .Loader {
    @include panel-2();

    position: relative;

    > * {
      position: absolute;
      top: 50%;
      left: 50%;
      transform: translate3d(-50%, -50%, 0);
    }
  }

  .isClientRendered &.isLoading img {
    height: 0;
    opacity: 0;
  }
}
