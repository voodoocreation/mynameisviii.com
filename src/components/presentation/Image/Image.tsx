import cn from "classnames";
import * as React from "react";

import Loader from "../Loader/Loader";

interface IProps {
  alt: string;
  caption?: string;
  className?: string;
  onClick?: () => void;
  onLoad?: () => void;
  src: string;
  title?: string;
}

interface IState {
  isRendered: boolean;
}

export default class Image extends React.Component<IProps, IState> {
  public readonly state = {
    isRendered: false
  };

  private imageRef: React.RefObject<HTMLImageElement> = React.createRef();

  public componentDidMount() {
    if (this.imageRef.current && this.imageRef.current.complete) {
      setTimeout(this.onLoad, 1);
    }
  }

  public render() {
    const {
      alt,
      caption,
      children,
      className,
      onClick,
      src,
      title,
      ...props
    } = this.props;
    const { isRendered } = this.state;

    const clickableProps = onClick
      ? {
          onClick,
          onKeyPress: this.onKeyPress,
          role: "button",
          tabIndex: 0
        }
      : undefined;

    return (
      <figure
        className={cn(
          "Image",
          className,
          { isLoading: !isRendered },
          { isRendered }
        )}
        {...clickableProps}
        {...props}
      >
        <div className="Image-wrapper">
          <img
            src={src}
            alt={alt}
            onLoad={this.onLoad}
            onError={this.onLoad}
            ref={this.imageRef}
            title={title}
          />
          {!this.state.isRendered ? <Loader /> : null}
          {children}
        </div>
        {caption ? <figcaption>{caption}</figcaption> : null}
      </figure>
    );
  }

  private onKeyPress = (event: React.KeyboardEvent<HTMLElement>) => {
    if (event.key === "Enter" && this.props.onClick) {
      this.props.onClick();
    }
  };

  private onLoad = () => {
    this.setState({
      isRendered: true
    });

    if (this.props.onLoad) {
      this.props.onLoad();
    }
  };
}
