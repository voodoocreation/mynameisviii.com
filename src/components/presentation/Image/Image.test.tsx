import { mount, render } from "enzyme";
import * as React from "react";

import Image from "./Image";

const g: any = global;

const setup = (fn: any, fromTestProps?: any) => {
  const props = {
    className: "TestImage",
    ...fromTestProps
  };

  return {
    actual: fn(<Image {...props} />),
    props
  };
};

describe("[presentation] <Image />", () => {
  beforeEach(() => {
    Object.defineProperty(g.Image.prototype, "complete", {
      value: false
    });
  });

  it("renders correctly with minimum props", () => {
    const { actual } = setup(render);

    expect(actual.find(".Loader")).toHaveLength(1);
    expect(actual).toMatchSnapshot();
  });

  it("triggers `onClick` prop when `Enter` key is pressed on focused image", () => {
    const { actual, props } = setup(mount, { onClick: jest.fn() });

    actual.simulate("keypress", { key: "Enter" });

    expect(props.onClick).toHaveBeenCalled();
  });

  it("doesn't trigger `onClick` prop when a key other than `Enter` is pressed on focused image", () => {
    const { actual, props } = setup(mount, { onClick: jest.fn() });

    actual.simulate("keypress", { key: "ArrowLeft" });

    expect(props.onClick).not.toHaveBeenCalled();
  });

  it("updates `isRendered` state and triggers `onLoad` prop after image has loaded", () => {
    const { actual, props } = setup(mount, {
      onLoad: jest.fn()
    });

    actual.find("img").simulate("load");

    expect(props.onLoad).toHaveBeenCalled();
    expect(actual.find("Loader")).toHaveLength(0);
    expect(actual.render()).toMatchSnapshot();
  });

  it("updates `isRendered` state when image has previously been loaded", () => {
    Object.defineProperty(g.Image.prototype, "complete", {
      value: true
    });

    const { actual } = setup(mount);

    actual.find("img").simulate("load");

    expect(actual.find("Loader")).toHaveLength(0);
    expect(actual.render()).toMatchSnapshot();
  });
});
