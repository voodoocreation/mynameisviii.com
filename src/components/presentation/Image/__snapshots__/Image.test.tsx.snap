// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[presentation] <Image /> renders correctly with minimum props 1`] = `
<figure
  class="Image TestImage isLoading"
>
  <div
    class="Image-wrapper"
  >
    <img />
    <span
      class="Loader"
    >
      <div
        class="Loader-animation"
      >
        <span />
        <span />
        <span />
      </div>
    </span>
  </div>
</figure>
`;

exports[`[presentation] <Image /> updates \`isRendered\` state and triggers \`onLoad\` prop after image has loaded 1`] = `
<figure
  class="Image TestImage isRendered"
>
  <div
    class="Image-wrapper"
  >
    <img />
  </div>
</figure>
`;

exports[`[presentation] <Image /> updates \`isRendered\` state when image has previously been loaded 1`] = `
<figure
  class="Image TestImage isRendered"
>
  <div
    class="Image-wrapper"
  >
    <img />
  </div>
</figure>
`;
